staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/dynamic_cafile_content.go

```diff
@@ -158,17 +159,51 @@ func (c *DynamicFileCAContent) Run(workers int, stopCh <-chan struct{}) {
 	// doesn't matter what workers say, only start one.
 	go wait.Until(c.runWorker, time.Second, stopCh)
 
-	// start timer that rechecks every minute, just in case.  this also serves to prime the controller quickly.
-	go wait.PollImmediateUntil(FileRefreshDuration, func() (bool, error) {
-		c.queue.Add(workItemKey)
-		return false, nil
-	}, stopCh)
-
-	// TODO this can be wired to an fsnotifier as well.
+	// start the loop that watches the CA file until stopCh is closed.
+	go wait.Until(func() {
+		if err := c.watchCAFile(stopCh); err != nil {
+			klog.ErrorS(err, "Failed to watch CA file, will retry later")
+		}
+	}, time.Minute, stopCh)
 
 	<-stopCh
 }
 
+func (c *DynamicFileCAContent) watchCAFile(stopCh <-chan struct{}) error {
+	w, err := fsnotify.NewWatcher()
+	if err != nil {
+		return fmt.Errorf("error creating fsnotify watcher: %v", err)
+	}
+	defer w.Close()
+
+	if err = w.Add(c.filename); err != nil {
+		return fmt.Errorf("error adding watch for file %s: %v", c.filename, err)
+	}
+	// Reload the content when it starts a new watch as the file may have changed.
+	c.queue.Add(workItemKey)
+
+	for {
+		select {
+		case e := <-w.Events:
+			c.queue.Add(workItemKey)
+			// The original watched file is removed or renamed, restarting a new watch.
+			if e.Op&(fsnotify.Remove|e.Op&fsnotify.Rename) > 0 {
+				if err := w.Remove(c.filename); err != nil {
+					klog.InfoS("Failed to remove file watch, it may have been deleted", "file", c.filename, "err", err)
+				}
+				if err := w.Add(c.filename); err != nil {
+					return fmt.Errorf("error adding watch for file %s: %v", c.filename, err)
+				}
+				c.queue.Add(workItemKey)
```

Comment:
I updated this a little to guarantee a resync will be triggered after any event is received and it will be executed after starting new watch if there is.